src/querystring.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{error::Error, Context};
+use http::StatusCode;
+use serde::Deserialize;
+
+/// An extension trait for `Context`, providing query string deserialization.
+pub trait ExtractQuery<'de> {
+    fn query<T: Deserialize<'de>>(&'de self) -> Result<T, Error>;
+}
+
+impl<'de, Data> ExtractQuery<'de> for Context<Data> {
+    fn query<T: Deserialize<'de>>(&'de self) -> Result<T, Error> {
+        let querystr = self.uri().query();
+
+        if querystr.is_none() {
+            return Err(Error::from(StatusCode::BAD_REQUEST));
+        }
+
+        let query = serde_urlencoded::from_str(querystr.unwrap())
+            .map_err(|_| Error::from(StatusCode::INTERNAL_SERVER_ERROR))?;
```

Comment:
Shouldn't this be a 400 error if we fail to parse?

Reply:
Agreed. I returned `BAD_REQUEST` earlier, not sure why I didn't here. It has been corrected.